File: content/08-heap-priority-queue/kth_largest_element.js
/**
 * Problem: Kth Largest Element in an Array
 * LeetCode: #215 (Medium) | NeetCode 150
 * 
 * Time Complexity: O(N log N) with sort or O(N log K) with min-heap
 * Space Complexity: O(1) or O(K)
 */

function findKthLargest(nums, k) {
    nums.sort((a, b) => b - a);
    return nums[k - 1];
}

if (require.main === module) {
    console.assert(findKthLargest([3, 2, 1, 5, 6, 4], 2) === 5, "Test 1 Failed");
    console.assert(findKthLargest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4) === 4, "Test 2 Failed");
    console.log("All JavaScript tests passed for Kth Largest Element!");
}

File: content/08-heap-priority-queue/kth_largest_element.py
"""
Problem: Kth Largest Element in an Array
LeetCode: #215 (Medium) | NeetCode 150

Given an integer array nums and an integer k, return the kth largest element in the array.

Time Complexity: O(N log K)
Space Complexity: O(K)
"""

import heapq

def find_kth_largest(nums: list[int], k: int) -> int:
    min_heap = []
    for num in nums:
        heapq.heappush(min_heap, num)
        if len(min_heap) > k:
            heapq.heappop(min_heap)
    return min_heap[0]

if __name__ == "__main__":
    assert find_kth_largest([3, 2, 1, 5, 6, 4], 2) == 5
    assert find_kth_largest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4) == 4
    print("All Python tests passed for Kth Largest Element!")
